### ISSUE TITLE
WaterTAP test failures with current IDAES main (1.13.dev) and 1.12.0 release

### ISSUE THREAD

**Issue by lbianchi-lbl** [watertap-org/watertap#287]
## Description

- Multiple tests in the WaterTAP test suite fail when using recent version(s) of IDAES:
  - Current `main` branch of the upstream repo IDAES/idaes-pse
  - 1.12.0rc1 (latest release candidate)
  - 1.12.0 (current release)
- The tests pass using the exact same WaterTAP version for previous versions of IDAES:
  - The latest tag of our fork at watertap-org/idaes-pse (which is the one being used as a requirement in the development version of WaterTAP)
  - 1.12.0rc0 (which is the one being used as a strict requirement in the 0.3.0 release)

A sample of the failures:

```log
____________________________________________________________ test_case2b_mid_conc_with_dissolution _____________________________________________________________

    @pytest.mark.component
    def test_case2b_mid_conc_with_dissolution():
>       model = model = run_case2(xA=1e-2, xB=1e-9, xAB=1e-2, scaling=True, rxn_config=reaction_log_solubility)

watertap/examples/chemistry/tests/test_solids.py:682: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
watertap/examples/chemistry/tests/test_solids.py:592: in run_case2
    model.fs.unit.initialize(optarg=solver.options, outlvl=idaeslog.DEBUG)
../idaes-pse/idaes/core/unit_model.py:628: in initialize
    flags = blk.control_volume.initialize(
../idaes-pse/idaes/core/control_volume0d.py:1367: in initialize
    in_flags = blk.properties_in.initialize(
../idaes-pse/idaes/generic_models/properties/core/generic/generic_property.py:1272: in initialize
    sum(blk[k].flow_mol_phase_comp_apparent[p, jj]
../idaes-pse/idaes/generic_models/properties/core/generic/generic_property.py:1272: in <genexpr>
    sum(blk[k].flow_mol_phase_comp_apparent[p, jj]
/opt/conda/envs/test-watertap-idaes-1.12/lib/python3.8/site-packages/pyomo/core/base/indexed_component.py:553: in __getitem__
    validated_index = self._validate_index(index)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyomo.core.base.var.IndexedVar object at 0x7f94436e48e0>, idx = ('Liq', 'AB')

    def _validate_index(self, idx):
        if not IndexedComponent._DEFAULT_INDEX_CHECKING_ENABLED:
            # Return whatever index was provided if the global flag dictates
            # that we should bypass all index checking and domain validation
            return idx
    
        # This is only called through __{get,set,del}item__, which has
        # already trapped unhashable objects.
        validated_idx = self._index.get(idx, _NotFound)
        if validated_idx is not _NotFound:
            # If the index is in the underlying index set, then return it
            #  Note: This check is potentially expensive (e.g., when the
            # indexing set is a complex set operation)!
            return validated_idx
    
        if idx.__class__ is IndexedComponent_slice:
            return idx
    
        if normalize_index.flatten:
            # Now we normalize the index and check again.  Usually,
            # indices will be already be normalized, so we defer the
            # "automatic" call to normalize_index until now for the
            # sake of efficiency.
            normalized_idx = normalize_index(idx)
            if normalized_idx is not idx:
                idx = normalized_idx
                if idx in self._data:
                    return idx
                if idx in self._index:
                    return idx
        # There is the chance that the index contains an Ellipsis,
        # so we should generate a slicer
        if idx is Ellipsis or idx.__class__ is tuple and Ellipsis in idx:
            return self._processUnhashableIndex(idx)
        #
        # Generate different errors, depending on the state of the index.
        #
        if not self.is_indexed():
            raise KeyError(
                "Cannot treat the scalar component '%s' "
                "as an indexed component" % ( self.name, ))
        #
        # Raise an exception
        #
>       raise KeyError(
            "Index '%s' is not valid for indexed component '%s'"
            % ( idx, self.name, ))
E       KeyError: "Index '('Liq', 'AB')' is not valid for indexed component 'fs.unit.control_volume.properties_in[0.0].flow_mol_phase_comp_apparent'"

/opt/conda/envs/test-watertap-idaes-1.12/lib/python3.8/site-packages/pyomo/core/base/indexed_component.py:778: KeyError
```

Summary of the entire test suite run:

```txt
=================================================================== short test summary info ====================================================================
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_1_no_dissolution - KeyError: "Index '('Liq', 'Ca(OH)2')' is not valid for...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_1_high_dissolution - KeyError: "Index '('Liq', 'Ca(OH)2')' is not valid f...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_1_mid_dissolution - KeyError: "Index '('Liq', 'Ca(OH)2')' is not valid fo...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_1_low_dissolution - KeyError: "Index '('Liq', 'Ca(OH)2')' is not valid fo...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_1_high_precipitation - KeyError: "Index '('Liq', 'Ca(OH)2')' is not valid...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_1_low_precipitation - KeyError: "Index '('Liq', 'Ca(OH)2')' is not valid ...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_2_do_nothing - KeyError: "Index '('Liq', 'CaCO3')' is not valid for index...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_2_seawater_no_ca - KeyError: "Index '('Liq', 'CaCO3')' is not valid for i...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_2_seawater_with_ca - KeyError: "Index '('Liq', 'CaCO3')' is not valid for...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_2_seawater_added_carbonates - KeyError: "Index '('Liq', 'CaCO3')' is not ...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_2_low_pH_no_precip - KeyError: "Index '('Liq', 'CaCO3')' is not valid for...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_2_ultra_high_ca - KeyError: "Index '('Liq', 'CaCO3')' is not valid for in...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_3_seawater_with_ca - KeyError: "Index '('Liq', 'CaCO3')' is not valid for...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_3_ultra_high_ca_forced_lime_precip - KeyError: "Index '('Liq', 'CaCO3')' ...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_3_ultra_faux_added_lime_and_ash - KeyError: "Index '('Liq', 'CaCO3')' is ...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_4_seawater_added_Fe_OH_PO4 - KeyError: "Index '('Liq', 'FePO4')' is not v...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_4_seawater_added_Fe_PO4 - KeyError: "Index '('Liq', 'FePO4')' is not vali...
FAILED watertap/examples/chemistry/tests/test_pH_dependent_solubility.py::test_case_4_seawater_added_Fe_only - KeyError: "Index '('Liq', 'FePO4')' is not val...
FAILED watertap/examples/chemistry/tests/test_recarbonation_process.py::TestCarbonationProcess::test_initialize_solver - KeyError: "Index '('Vap', 'H2CO3')' ...
FAILED watertap/examples/chemistry/tests/test_recarbonation_process.py::TestCarbonationProcess::test_solve_equilibrium - RuntimeError: Error in AMPL evaluation.
FAILED watertap/examples/chemistry/tests/test_recarbonation_process.py::TestCarbonationProcess::test_solution_equilibrium - assert 298 ± 3.0e-03 == 300
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case1_low_conc_no_precipitation - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case1_mid_conc_no_precipitation - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case1_high_conc_with_precipitation - KeyError: "Index '('Liq', 'AB')' is not valid for indexed ...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case1_low_conc_with_dissolution - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case1_mid_conc_with_dissolution - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case1_high_conc_for_all - KeyError: "Index '('Liq', 'AB')' is not valid for indexed component '...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2_low_conc_no_precipitation - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2_mid_conc_no_precipitation - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2_high_conc_with_precipitation - KeyError: "Index '('Liq', 'AB')' is not valid for indexed ...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2_low_conc_with_dissolution - KeyError: "Index '('Liq', 'AB')' is not valid for indexed com...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2a_mid_conc_with_dissolution - KeyError: "Index '('Liq', 'AB')' is not valid for indexed co...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2b_mid_conc_with_dissolution - KeyError: "Index '('Liq', 'AB')' is not valid for indexed co...
FAILED watertap/examples/chemistry/tests/test_solids.py::test_case2_high_conc_for_all - KeyError: "Index '('Liq', 'AB')' is not valid for indexed component '...
FAILED watertap/examples/chemistry/tests/test_tutorials.py::TestTutorials::test_tutorials - nbclient.exceptions.CellExecutionError: An error occurred while e...
FAILED watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/test_posttreatment.py::test_ideal_naocl_mixer - ZeroDivisionError: fl...
FAILED watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/test_posttreatment.py::test_ideal_naocl_chlorination_full_block - Zer...
FAILED watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/test_posttreatment.py::test_addition_of_translator - ZeroDivisionErro...
FAILED watertap/property_models/tests/test_NaCl_prop_pack.py::TestNaClProperty_idaes::test_initialize_failure - Failed: DID NOT RAISE <class 'idaes.core.util...
FAILED watertap/property_models/tests/test_seawater_prop_pack.py::TestSeawaterProperty_idaes::test_initialize_failure - Failed: DID NOT RAISE <class 'idaes.c...
==================================================== 40 failed, 502 passed, 8 skipped in 309.54s (0:05:09) =====================================================
```

## Steps to reproduce

(**NOTE**: I've used slightly different steps so these might not work as-is. The important thing is to make sure that the `idaes` package being used by WaterTAP comes from the local `idaes-pse` repo installed in editable mode)

- Create a new environment and install the latest version of WaterTAP (either the current `main` or the 0.3.0 release)
- Clone the IDAES/idaes-pse repo in a separate directory, enter the directory, and install `idaes-pse` in editable mode: `pip install -e --ignore-installed .`
- Run pytest with the appropriate options to ensure that the WaterTAP test suite, and not the IDAES test suite, is being run, e.g. `pytest --pyargs watertap -c /dev/null` (or `-c NUL` on Windows)
  - I've used a subset of the test suite as a representative sample to speed things up by adding the `-k test_solids` flag
- Check out another commit or tag in the IDAES repository and repeat the test

## Possible cause

Using `git bisect` in the IDAES repository (basically automating the process described above), the first failing commit seems to be [8ce5d82d](https://github.com/IDAES/idaes-pse/commit/8ce5d82d5e79c6bacc5af9fd2c5d6ddf8c3f060a), corresponding to PR IDAES/idaes-pse#622:

```txt
commit 8ce5d82d5e79c6bacc5af9fd2c5d6ddf8c3f060a
Author: Andrew Lee <andrew.lee@netl.doe.gov>
Date:   Thu Dec 9 16:48:15 2021 -0500

    Fixing issues that were causing Pyomo to emit warnings about setting variables outside of bounds

 .../technical_specs/core/util/model_serializer.rst |  2 +-
 .../properties/core/generic/generic_property.py    | 26 +++++++++++++++++-----
 .../column_models/tests/test_solvent_condenser.py  |  4 ++--
 3 files changed, 23 insertions(+), 9 deletions(-)
```

The `git bisect` commands I ran in the IDAES repo:

```sh
git bisect start
git checkout main
git bisect bad
git checkout 1.12.0rc0
git bisect good
git bisect run pytest --pyargs watertap -k test_solids -qq -c /dev/null --tb=no
```

**Comment by andrewlee94** [comment]
Two of these failures are due to the addition of a new Exception in IDAES for `initialization` methods, which are now expected in the initialization tester but not yet implemented in watertap.

The issue above however is unexpected, and I am not immediately sure what caused it.

**Comment by andrewlee94** [comment]
I have found the issue (or at least the first one) that was resulting in the failures @lbianchi-lbl showed above - it was due to sparse phase-component sets which were not being checked.

**Comment by aladshaw3** [comment]
Issues resolved from #289
